Yosemite/Yosemite/Stores/ProductStore.swift

```diff
@@ -512,6 +513,31 @@ private extension ProductStore {
         }
         storageProduct.addToTags(NSOrderedSet(array: storageTags))
     }
+
+    /// Updates, inserts, or prunes the provided StorageProduct's downloadable files using the provided read-only Product's downloadable files
+    ///
+    func handleProductDownloadableFiles(_ readOnlyProduct: Networking.Product, _ storageProduct: Storage.Product, _ storage: StorageType) {
+
+        // Removes all the downloadable files first.
+        storageProduct.downloadableFilesArray.forEach { existingStorageDownloadableFile in
+            storage.deleteObject(existingStorageDownloadableFile)
+            storageProduct.removeFromDownloads(existingStorageDownloadableFile)
+        }
+
+        var storageDownloads = [StorageProductDownload]()
+        for readOnlyDownloadableFile in readOnlyProduct.downloads {
+
+            if let existingStorageDownloadableFile = storage.loadProductDownloadableFile(downloadableFileID: readOnlyDownloadableFile.downloadID) {
+                existingStorageDownloadableFile.update(with: readOnlyDownloadableFile)
+                storageDownloads.append(existingStorageDownloadableFile)
+            } else {
+                let newStorageDownloadableFile = storage.insertNewObject(ofType: Storage.ProductDownload.self)
+                newStorageDownloadableFile.update(with: readOnlyDownloadableFile)
+                storageDownloads.append(newStorageDownloadableFile)
+            }
+        }
+        storageProduct.addToDownloads(NSOrderedSet(array: storageDownloads))
```

Comment:
What do you think about extracting from line 522 to line 525 and from line 528 to line 538 into their own private methods? That would increase the readability of this chunk!

Reply:
I think that would be a great idea and also readability would improve a lot. 
But I am not sure on one point. All other methods on this class that are performing similar types of tasks (as an example `handleProductTags` , `handleProductImages` , `handleProductDefaultAttributes` , `handleProductAttributes` etc) are designed in the similar fashion as it is currently/above. Taking that into consideration what do you suggest?

@pmusolino , would you please share your thoughts on this?

Reply:
I bring that up because I had to read the method several times in order to understand what it was doing. Having smaller helper methods with meaningful names would have turned that into a single read. I think we can always improve what we have been doing so far, but I'll let @pmusolino to decide! :-)

Reply:
I'd say that, just for not having discrepancies inside the same class, we can split it, adding some comments that better describe every pass of the method. In the future, we can think about some generic ways for better split things. What do you think?

Reply:
I agree with both of you. I have splited that into 2 smaller methods.